### DISCUSSION TITLE
How to apply overlay but while maintaining the transparency

### DISCUSSION THREAD

**Discussion by LuukOV** [ImageMagick/ImageMagick#6574] [Help]
### ImageMagick version

7.1.0-55

### Operating system, version and so on

Windows 10

### Description

I have this image that I want to make lighter using the following script:

`magick image.png -size 1000x1000 canvas:white -compose overlay -composite output.png`

![screenshot - 2023-08-18T101405 575](https://github.com/ImageMagick/ImageMagick/assets/27200792/d52fb52c-3f8e-47a8-9f52-2d2cf018052e)

This results in this:

![overlayed](https://github.com/ImageMagick/ImageMagick/assets/27200792/bc2579d0-09a6-4548-9335-d3ebcad6bc49)

Which has the color result I want. However, now I also want it to maintain it's transparency. So that if I would put the end result  in photoshop and put it infront of a white layer, it looks the same as the 2nd image. Can anyone help me to get this result?

**Comment by snibgo** [comment]
> So that if I would put the end result in photoshop and put it infront of a white layer...

What does that mean? I don't use Photoshop. I assume it means "composite the result over a white image of the same size".

I don't think there is an exact solution. An approximate solution is to multiply the colour channels by two. I named the first of your images `redCirc.png`.
```
magick redCirc.png -channel RGB -evaluate Multiply 2 +channel xa.png
```
When xa.png is composited over white, the result has RMSE of 0.36% from the desired result, which is very close.

xa.png has essentially two colours in the red circle: sRGB(99.6078%,0,0) and sRGB(98.8235%,0,0). So a slightly worse (but close) solution would use pure red.

**Comment by fmw42** [comment]
Sorry, I do not understand what you want?  Do you want us to tell you how to use Photoshop? Or something with Imagemagick only?  Can you show an example of what you want?

**Comment by LuukOV** [comment]
Let me try to explain better. I have the following source image:
![image](https://github.com/ImageMagick/ImageMagick/assets/27200792/449115d5-81ac-44ea-89fc-0dac6a3381bf)

I also have the following image:
![image (46)](https://github.com/ImageMagick/ImageMagick/assets/27200792/01c6a06d-fc12-4a5b-be94-4fa0ab104cdc)

This is the resulting color I want (the 2nd image). But I need it to maintain transparency. The reason that I want it to maintain transparency is that I want to be able to put it in front of other images but make it have the colors of the 2nd image.

So this:
![image](https://github.com/ImageMagick/ImageMagick/assets/27200792/9533834f-b587-4e12-967d-943280c43a13)
Should look like the 2nd image.

I need this for an automated process, so I'm trying to create an imagemagick script that can give me this result but I'm struggling to make it consistent.

For reference I also have the following images:
(Source)
![screenshot - 2023-08-22T084500 832](https://github.com/ImageMagick/ImageMagick/assets/27200792/f27f79a8-8d11-47e8-8eec-2e99ce014ddb)

(Desired color result)
![screenshot - 2023-08-22T084455 577](https://github.com/ImageMagick/ImageMagick/assets/27200792/110371bb-6bc8-4386-802b-d46b26cdb9f7)

My assumption is that only the transparent pixels should be mixed with a level of white.

**Reply by fmw42** [reply]
Which red dot image is your input and which is the output?  If you give me those, I can try to reconstruct in Photoshop and then convert that process to Imagemagick. Also tell me your Photoshop steps.  Please be very clear in your explanation as your past posts have been confusing to me.

**Reply by LuukOV** [reply]
Look at my reply to snibgo

**Comment by snibgo** [comment]
> This is the resulting color I want (the 2nd image). But I need it to maintain transparency. The reason that I want it to maintain transparency is that I want to be able to put it in front of other images but make it have the colors of the 2nd image.

This confuses me. In your first image, which I have named `redCirc.png`, almost all pixels have some transparency. (4 white pixels are opaque.) Most pixels are fully transparent. The red pixels are partially transparent.

I thought you wanted an image that has exactly the same transparency as `redCirc.png`. So partially transparent pixels should remain partially transparent, with the same transparency. Is that correct? But you say:
> I want to be able to put it in front of other images but make it have the colors of the 2nd image.

If you put semi-transparent red "in front of" a strong colour such as blue, the result will be a reddish blue. It won't have the "colors of the 2nd image".

If you want to keep the same red colour no matter what colour the image is "put in front of", then the red pixels must be fully opaque. Is that what you want?

**Reply by LuukOV** [reply]
> I want to be able to put it in front of other images but make it have the colors of the 2nd image.

In this case if it'd be in front of a white background, like I showcased in the photoshop example (photoshop is just the tool I used in this case, might as well have been GIMP or Paint.net etc).

As you can see, when I put my input image in front of a white background, the red is darker than my desired output image. My goal is for me to manipulate the input image in such a way that when I perform the same as I've done here:
![image](https://github.com/ImageMagick/ImageMagick/assets/27200792/8053a166-f7c6-4cf9-b087-1aaaecba9397)

It looks like this:
![image](https://github.com/ImageMagick/ImageMagick/assets/27200792/b5b61a52-d4ab-4f15-b562-790ec57ef759)

I've also provided the image of the glass, so that the script can be tested on different images as well. The first image of the glass needs to look like the second image of the glass in the same way as the red sphere. (when put in front of a white background)

> If you want to keep the same red colour no matter what colour the image is "put in front of", then the red pixels must be fully opaque. Is that what you want?

No

As some background information, I'm getting these image from a WebEngine called Babylon.js. I can take screenshots from this engine by either blending with an opaque white, transparent white or transparent black. The "result" images (output) I have are from blending with an opaque white. I however need to maintain the transparency for later editing, so I'm trying to manipulate my "transparent black" (input, redCirc.png) image into a way that it looks like the opaque white image (output) while maintaining the transparency of the input(redCirc.png). Most likely we just need to manipulate the RGB values of pixels that contain any level of transparency and turn them lighter. I just don't know what way I should blend them to get them to look like the output image.

**Comment by snibgo** [comment]
3 days ago I suggested a result `xa.png` for the red circle image. As far as I can see from your latest post, that is the result you want. If it isn't, please say what is wrong with it.

The command isn't a general solution, but I think it is the solution for the red circle example.
